openvr_ros_driver/src/openvr_driver.cpp

```diff
@@ -0,0 +1,360 @@
+
+    case (vr::VREvent_Quit) :
+    {
+      char buf[1024];
+      sprintf(buf, "Received SteamVR Quit (%d)", vr::VREvent_Quit);
+      printf("%s\n", buf);
+
+      return false;
+    }
+
+    case (vr::VREvent_ProcessQuit) :
+    {
+      char buf[1024];
+      sprintf(buf, "SteamVR Quit Process (%d)", vr::VREvent_ProcessQuit);
+      printf("%s\n", buf);
+
+      return false;
+    }
+
+    case (vr::VREvent_QuitAborted_UserPrompt) :
+    {
+      char buf[1024];
+      sprintf(buf, "SteamVR Quit Aborted UserPrompt (%d)", vr::VREvent_QuitAborted_UserPrompt);
+      printf("%s\n", buf);
+
+      return false;
+    }
+
+    case (vr::VREvent_QuitAcknowledged) :
+    {
+      char buf[1024];
+      sprintf(buf, "SteamVR Quit Acknowledged (%d)", vr::VREvent_QuitAcknowledged);
+      printf("%s\n", buf);
+
+      return false;
+    }
+
+    case (vr::VREvent_TrackedDeviceRoleChanged) :
+    {
+
+      char buf[1024];
+      sprintf(buf, "TrackedDeviceRoleChanged: %d", event.trackedDeviceIndex);
+      printf("%s\n", buf);
+      break;
+    }
+
+    case (vr::VREvent_TrackedDeviceUserInteractionStarted) :
+    {
+      char buf[1024];
+      sprintf(buf, "TrackedDeviceUserInteractionStarted: %d", event.trackedDeviceIndex);
+      printf("%s\n", buf);
+      break;
+    }
+
+    default:
+      if (event.eventType == vr::VREvent_None) break;
+
+      char buf[1024];
+      sprintf(buf, "Event: %s (%d)", vr::VRSystem()->GetEventTypeNameFromEnum((vr::EVREventType) event.eventType), event.eventType);
+      printf("%s\n", buf);
+      break;
+  }
+
+  return true;
+}
+
+// Get the quaternion representing the rotation
+vr::HmdQuaternion_t OpenVRDriver::getRotation(vr::HmdMatrix34_t matrix) {
+  vr::HmdQuaternion_t q;
+
+  q.w = sqrt(fmax(0, 1 + matrix.m[0][0] + matrix.m[1][1] + matrix.m[2][2])) / 2;
+  q.x = sqrt(fmax(0, 1 + matrix.m[0][0] - matrix.m[1][1] - matrix.m[2][2])) / 2;
+  q.y = sqrt(fmax(0, 1 - matrix.m[0][0] + matrix.m[1][1] - matrix.m[2][2])) / 2;
+  q.z = sqrt(fmax(0, 1 - matrix.m[0][0] - matrix.m[1][1] + matrix.m[2][2])) / 2;
+  q.x = copysign(q.x, matrix.m[2][1] - matrix.m[1][2]);
+  q.y = copysign(q.y, matrix.m[0][2] - matrix.m[2][0]);
+  q.z = copysign(q.z, matrix.m[1][0] - matrix.m[0][1]);
+  return q;
+}
+
+// Get the vector representing the position
+vr::HmdVector3_t OpenVRDriver::getPosition(vr::HmdMatrix34_t matrix) {
+  vr::HmdVector3_t vector;
+
+  vector.v[0] = matrix.m[0][3];
+  vector.v[1] = matrix.m[1][3];
+  vector.v[2] = matrix.m[2][3];
+
+  return vector;
+}
+
+void OpenVRDriver::publishTrackingData() {
+  vive_msgs::ViveSystem msg;
+
+  // Process SteamVR device states
+  for (vr::TrackedDeviceIndex_t unDevice = 0; unDevice < vr::k_unMaxTrackedDeviceCount; unDevice++)
+  {
+    // if not connected just skip the rest of the routine
+    if (!m_pHMD->IsTrackedDeviceConnected(unDevice))
+      continue;
+
+    vr::TrackedDevicePose_t trackedDevicePose;
+    vr::TrackedDevicePose_t *devicePose = &trackedDevicePose;
```

Comment:
No need to use both the pointer and object. Further comments will adjust code to just use object `trackedDevicePose`, not the pointer